### PR TITLE
pyflyte package command

### PR DESCRIPTION
**Pull request by kumare3**
Signed-off-by: Ketan Umare <ketan.umare@gmail.com>

# TL;DR
Introduces new pyflyte package command. This command is intended to replace all `serialize, fast-serialize` commands and is intended to clean up the UX of packaging flyte apps.

`pyflyte --pkgs myapp.workflows package --image myapp:03eccc1cf101adbd8c4734dba865d3fdeb720aa7 -f --fast`

This produces a single tar-ball that has all the serialized workflows, tasks and launchplans.
The output of this command is intended to be used by `flytectl register`

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

### CODE REVIEW

File: flytekit/clis/sdk_in_container/package.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    Validates the image to match the standard format. Also validates that only one default image
+    is provided. a default image, is one that is specified as
+      default=img or just img. All other images should be provided with a name, in the format
+      name=img
+    """
+    default_image = None
+    images = []
+    for v in values:
+        if "=" in v:
+            vals = v.split("=")
```
Comment (EngHabu) [inline flytekit/clis/sdk_in_container/package.py:36]:
```suggestion
            vals = v.split("=", maxsplit=1)
```

File: flytekit/clis/sdk_in_container/package.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            if len(vals) > 2:
+                raise click.BadParameter(f"Bad value received [{v}] for param {param}, more than 1 `=` is not allowed")
```
Comment (EngHabu) [inline flytekit/clis/sdk_in_container/package.py:38]:
```suggestion
```

File: flytekit/clis/sdk_in_container/package.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            default_image = img
+        else:
+            images.append(img)
+
+    return ImageConfig(default_image, images)
+
+
+def get_registrable_entities(ctx: context_manager.FlyteContext) -> typing.List:
+    """
+    Returns all entities that can be serialized and should be sent over to Flyte backend. This will filter any local entities
```
Comment (EngHabu) [inline flytekit/clis/sdk_in_container/package.py:59]: What are "local entities" ?

File: flytekit/clis/sdk_in_container/package.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            )
+        if img.name == _DEFAULT_IMAGE_NAME:
+            default_image = img
+        else:
+            images.append(img)
+
+    return ImageConfig(default_image, images)
+
+
+def get_registrable_entities(ctx: context_manager.FlyteContext) -> typing.List:
```
Comment (EngHabu) [inline flytekit/clis/sdk_in_container/package.py:57]: It feels like this function must exist elsewhere... right?

File: flytekit/clis/sdk_in_container/package.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+from flytekit.core.base_task import PythonTask
+from flytekit.core.context_manager import ImageConfig, look_up_image_info
+from flytekit.core.workflow import WorkflowBase
+from flytekit.tools import fast_registration, module_loader
+
+_DEFAULT_IMAGE_NAME = "default"
+_DEFAULT_RUNTIME_PYTHON_INTERPRETER = "/opt/venv/bin/python3"
+
+
+def validate_image(ctx: typing.Any, param: str, values: tuple) -> ImageConfig:
```
Comment (EngHabu) [inline flytekit/clis/sdk_in_container/package.py:25]: It feels like this function exist elsewhere, right?

Comment (katrogan) [reply flytekit/clis/sdk_in_container/package.py:25]: Soren added it! https://github.com/flyteorg/flytekit/pull/394/files#diff-ded0a6913453e62ae1a80623d886c46931197ef4d206f0e10155475a8f356df0R68

Comment (katrogan) [reply flytekit/clis/sdk_in_container/package.py:25]: https://github.com/flyteorg/flytekit/pull/394/files#diff-ded0a6913453e62ae1a80623d886c46931197ef4d206f0e10155475a8f356df0R68

File: flytekit/clis/sdk_in_container/package.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                translator.get_serializable(new_api_serializable_entities, ctx.serialization_settings, lp)
+
+    new_api_model_values = list(new_api_serializable_entities.values())
+    entities_to_be_serialized = list(filter(serialize._should_register_with_admin, new_api_model_values))
+    return [v.to_flyte_idl() for v in entities_to_be_serialized]
+
+
+def persist_registrable_entities(entities: typing.List, folder: str):
+    """
+    For protobuf serializable list of entities, writes a file with the name if the entity and
```
Comment (katrogan) [inline flytekit/clis/sdk_in_container/package.py:79]:
```suggestion
    For protobuf serializable list of entities, writes a file with the name of the entity, entity type and
```

File: flytekit/clis/sdk_in_container/package.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    help="Filesystem path to where the code is copied into within the Dockerfile. look for `COPY . /root` like command.",
+)
+@click.pass_context
+def package(ctx, image_config, source, output, force, fast, in_container_source_path, python_interpreter):
+    """
+    This command produces a Flyte backend registrable package of all entities in Flyte.
+    For tasks, one pb file is produced for each task, representing one TaskTemplate object.
+    For workflows, one pb file is produced for each workflow, representing a WorkflowClosure object.  The closure
+        object contains the WorkflowTemplate, along with the relevant tasks for that workflow.
+        This serialization step will set the URN of the tasks to the fully qualified name of the task function.
```
Comment (katrogan) [inline flytekit/clis/sdk_in_container/package.py:175]: URN or just name?

File: flytekit/clis/sdk_in_container/package.py (hunk 1 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/package.py:36]: Are you suggesting that if  there are more than 1 `=` we should just take it and let the image library figure that out?

File: flytekit/clis/sdk_in_container/package.py (hunk 3 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/package.py:59]: aah local here means flytekit only, let me change that

File: flytekit/clis/sdk_in_container/package.py (hunk 4 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/package.py:57]: it does exist as part of the old code, this is made into a function.

File: flytekit/clis/sdk_in_container/package.py (hunk 5 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/package.py:25]: sadly no. It does use the actual parse image function that exists

File: flytekit/clis/sdk_in_container/package.py (hunk 4 above)
Comment (katrogan) [reply flytekit/clis/sdk_in_container/package.py:57]: it does, in serialize - but are we going to deprecate that as we move over to package?

File: flytekit/clis/sdk_in_container/pyflyte.py [hunk 8]
```diff
@@ -39,7 +50,7 @@
     "--insecure",
     required=False,
     type=bool,
-    help="Do not use SSL to connect to Admin",
+    help="Disable SSL to connect to Flyte backend.",
```
Comment (katrogan) [inline flytekit/clis/sdk_in_container/pyflyte.py:53]:
```suggestion
    help="Disable SSL when connecting to the Flyte backend.",
```

File: flytekit/clis/sdk_in_container/package.py (hunk 2 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/package.py:38]: done

File: flytekit/clis/sdk_in_container/package.py (hunk 1 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/package.py:36]: done

File: flytekit/clis/sdk_in_container/package.py (hunk 4 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/package.py:57]: ok i refactored old code

File: flytekit/clis/sdk_in_container/package.py (hunk 7 above)
Comment (kumare3) [reply flytekit/clis/sdk_in_container/package.py:118]: done

**Comment by kumare3** [comment]
PTAL @katrogan @EngHabu

**Review by katrogan — APPROVED** [review]
